rfd/0106-machine-id-anonymous-telemetry.md

```diff
@@ -0,0 +1,153 @@
+
+Event collection and submission will be started concurrently to the `tbot`
+functionality, and should not impede the primary function of `tbot`. In the
+case of failure, a warning message should be omitted. The routine will read in
+the configuration of `tbot`, extract the relevant values and encode these within
+an event protobuf.
+
+### Event submission
+
+Events will be submitted directly to an unauthenticated `tbot` specific endpoint
+of the `prehog` service:
+
+```protobuf
+syntax = "proto3";
+
+package prehog.v1alpha;
+
+import "google/protobuf/timestamp.proto";
+
+message SubmitTbotEventRequest {
+  // optional, will default to the ingest time if unset
+  google.protobuf.Timestamp timestamp = 1;
+
+  oneof event {
+    // See the events section for the fields included within startup.
+    TbotStartupEvent startup = 2;
+  }
+}
+
+message SubmitTbotEventResponse {}
+
+service TbotReportingService {
+  rpc SubmitTbotEvent(SubmitTbotEventRequest) returns (SubmitTbotEventResponse) {}
+}
+```
+
+As only a single event will be submitted, batching is not of concern.
+
+### Event storage
+
+Events received by `prehog` will be encoded in a PostHog compatible format
+and then submitted to PostHog.
+
+Tbot events will be share the same project as clusters and website events.
+
+### Consent
+
+Users will explicitly opt-in to anonymous usage telemetry with an environment
+variable, e.g:
+
+```sh
+# Send anonymous telemetry on startup about the usage of tbot in order
+# to help us understand what parts of Machine ID to improve.
+# Find out more at goteleport.com/awesome-docs-page
+export TBOT_ANONYMOUS_TELEMETRY=1
+tbot start --snip--
+```
+
+For helpers like GitHub actions, this will be another option the user can
+provide, e.g:
+
+```yaml
+steps:
+  - name: Install Teleport
+    uses: teleport-actions/setup@v1
+    with:
+      version: 11.0.3
+  - name: Authorize against Teleport
+    id: auth
+    uses: teleport-actions/auth@v1
+    with:
+      # Send anonymous telemetry on startup about the usage of tbot in order
+      # to help us understand what parts of Machine ID to improve.
+      # Find out more at goteleport.com/awesome-docs-page
+      anonymous-telemetry: true
+```
+
+Examples provided in the documentation will be updated to include this value
+to encourage opt-in, but will make clear that this parameter can be removed.
+
+A documentation page will be created that will explain what telemetry is
+collected and where this telemetry is stored.
+
+When telemetry is enabled, a log message will be output on startup stating that
+telemetry is being collected and linking to the relevant documentation.
+
+When telemetry is not enabled, a log message will be output on startup
+informing the user that Telemetry is not enabled and linking to the relevant
+documentation.
+
+### Anonymization
+
+The events will contain no properties that identify a user, `tbot` or Teleport
+cluster. Therefore, no additional anonymization is required at this time.
```

Comment:
Thinking about the PostHog encoding of the event, what would the "distinct ID" for the event be? Do we just generate a UUID on the fly for each event? Can we use some stable identifier for continuous mode, or maybe generate a dedicated one that's persisted in the tbot state? It would also help to avoid counting the same tbot multiple times, unless `tbot.start` is specifically for first starts.

Reply:
So for continuous mode, there will still be only a single event emitted on start. I'd be more than happy to just pursue generating a unique UUID on the fly if events require one.

We could introduce some sort of persistent identifier for a bot instance, but I'm not convinced it's useful to track a bot over its lifetime when Machine ID is often deployed into ephemeral environments with no ability to persist state. My gut feeling is that questions that this sort of thing would answer will be  better answered by the server-side metrics we'll look to introduce in Q2.

Reply:
I've added a `distinct_id`, which we'll generate (for now) for each event submission on start. If we end up adding multiple events in future, this ID can be used across all of them within a session.